src/material-experimental/mdc-form-field/form-field.ts

```diff
@@ -0,0 +1,462 @@
+    // in "ngDoCheck" based on the form-field control state. The following events
+    // need to be handled: focus, blur. We do not handle the "input" event since
+    // that one is only needed for the text-field character count, which we do
+    // not implement as part of the form-field, but should be implemented manually
+    // by consumers using template bindings.
+    registerInputInteractionHandler: () => {},
+    deregisterInputInteractionHandler: () => {},
+
+    // We do not have a reference to the native input since we work with abstract form field
+    // controls. MDC needs a reference to the native input optionally to handle character
+    // counting and value updating. These are both things we do not handle from within the
+    // form-field, so we can just return null.
+    getNativeInput: () => null,
+
+    // This method will never be called since we do not have the ability to add event listeners
+    // to the native input. This is because the form control is not necessarily an input, and
+    // the form field deals with abstract form controls of any type.
+    setLineRippleTransformOrigin: () => {},
+
+    // The foundation tries to register click and keyboard events on the form-field to figure out
+    // if the input value changes through user interaction. Based on that, the foundation tries
+    // to focus the input. Since we do not handle the input value as part of the form-field, nor
+    // it's guaranteed to be an input (see adapter methods above), this is a noop.
+    deregisterTextFieldInteractionHandler: () => {},
+    registerTextFieldInteractionHandler: () => {},
+
+    // The foundation tries to setup a "MutationObserver" in order to watch for attributes
+    // like "maxlength" or "pattern" to change. The foundation will update the validity state
+    // based on that. We do not need this logic since we handle the validity through the
+    // abstract form control instance.
+    deregisterValidationAttributeChangeHandler: () => {},
+    registerValidationAttributeChangeHandler: () => null as any,
+  };
+
+  constructor(private _elementRef: ElementRef,
+              private _changeDetectorRef: ChangeDetectorRef,
+              @Optional() @Inject(MAT_FORM_FIELD_DEFAULT_OPTIONS)
+              private _defaults?: MatFormFieldDefaultOptions,
+              @Optional() @Inject(MAT_LABEL_GLOBAL_OPTIONS) private _labelOptions?: LabelOptions,
+              @Optional() @Inject(ANIMATION_MODULE_TYPE) _animationMode?: string) {
+    this._animationsEnabled = _animationMode !== 'NoopAnimations';
+
+    if (_defaults && _defaults.appearance) {
+      this.appearance = _defaults.appearance;
+    } else if (_defaults && _defaults.hideRequiredMarker) {
+      this.hideRequiredMarker = true;
+    }
+  }
+
+  ngAfterViewInit() {
+    this._foundation = new MDCTextFieldFoundation(this._adapter);
+
+    Object.defineProperty(this._foundation, 'shouldFloat', {
```

Comment:
Add a comment for what this is doing?

Reply:
Yep, I was under the impression I added one.. done